### PR TITLE
TDL-17694-Added the missing test cases

### PR DESCRIPTION
**Pull request by umangagrawal-crest**
# Description of change
TDL-17694: Added missing test case

Added missing test cases for the below test files

- Automatic Fields Test
Verified that all replicated records have unique primary key values.

- All Fields Test
Verify no unexpected streams were replicated
Verify that all fields for each stream are replicated
Verify that more than just the automatic fields are replicated for each stream.

- Discovery Test
Verify there are no duplicate/conflicting metadata entries.
Verify stream names follow naming convention (lowercase alphas and underscores). --> This test case is not written as the stream names are Audience Overview, Audience Technology. Also, customers can create custom reports with any name

**NOTE**:  

- Bookmark tests are not modified as the PR (https://github.com/singer-io/tap-google-analytics/pull/67) is already raised as a part of https://jira.talendforge.org/browse/TDL-16943 
- Start date tests are not modified as no test cases are missing.
- Pagination tests are not uncommented as it is blocked on https://stitchdata.atlassian.net/browse/SRCE-5065 
- For All fields tests test cases are skipped due to open bug and API limitation 

# QA steps

 
# Risks
 No risks
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/base.py [hunk 1]
```diff
@@ -390,12 +390,12 @@ def expected_default_fields():
             },
             "Audience Overview": {
                 "ga:users", "ga:newUsers", "ga:sessions", "ga:sessionsPerUser", "ga:pageviews",
-                "ga:pageviewsPerSession", "ga:avgSessionDuration", "ga:bounceRate", "ga:date",
+                "ga:pageviewsPerSession", "ga:avgSessionDuration", "ga:bounceRate", "ga:date",'ga:month','ga:operatingSystem','ga:language','ga:hour','ga:browser','ga:year','ga:country','ga:city'
```
Comment (umangagrawal-crest) [inline tests/base.py:393]: Added the fields for all fields test case

**Review by dbshah1212 — APPROVED** [review]

**Comment by RushT007** [comment]
All fields test is skipping few streams due to tap limitations. Instead of skipping we should add separate test to test these streams with defined selection rules.

    - Few streams 200+ fields so testing all fields combinations will be time consuming 
    - Hardcoding field selection may not give complete coverage 

So we can randomise field selections and run the test. Note that, filed exclusion rules may impose some challenges.

**Review by RushT007 — CHANGES_REQUESTED** [review]

File: tests/test_google_analytics_all_fields.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    SKIP_STREAMS = {'Ecommerce Overview','Test Report 1', 'Audience Technology', 'Behavior Overview',}
+
+    def name(self):
+        return "tap_tester_google_analytics_all_fields_test"
+
+    def test_run(self):
+        """
+        • Verify no unexpected streams were replicated
+        • Verify that more than just the automatic fields are replicated for each stream. 
+        • verify all fields for each stream are replicated
```
[the remark this answers is not in the document]
Comment (umangagrawal-crest) [reply tests/test_google_analytics_all_fields.py]: Updated the comment.

File: tests/test_google_analytics_all_fields.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                # Verify that more than just the automatic fields are replicated for each stream.
+                self.assertTrue(expected_automatic_keys.issubset(
+                    expected_all_keys), msg='{} is not in "expected_all_keys"'.format(expected_automatic_keys-expected_all_keys))
+
+                messages = synced_records.get(stream)
+                # collect actual values
+                actual_all_keys = set()
+                for message in messages['messages']:
+                    if message['action'] == 'upsert':
+                        actual_all_keys.update(message['data'].keys())
```
[the remark this answers is not in the document]
Comment (umangagrawal-crest) [reply tests/test_google_analytics_all_fields.py:81]: Tried list comprehension but to achieve it, we need to import dependency of itertools. 
`actual_keys = [message['data'].keys() for message in messages['messages'] if message['action'] == 'upsert']
actual_all_keys = set(itertools.chain(*actual_keys))`
As discussed on slack, no need of changing existing implementation.

**Comment by umangagrawal-crest** [comment]
> All fields test is skipping few streams due to tap limitations. Instead of skipping we should add separate test to test these streams with defined selection rules.
> 
> ```
> - Few streams 200+ fields so testing all fields combinations will be time consuming 
> - Hardcoding field selection may not give complete coverage 
> ```
> 
> So we can randomise field selections and run the test. Note that, filed exclusion rules may impose some challenges.

@RushT007 As per your suggestion, added file which randomly selects fields for Audience Technology stream

File: tests/test_google_analytics_dynamic_dimension_metric.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    @staticmethod
+    def name():
+        return "tap_tester_google_analytics_dynamic_dimension_metric_field_test"
+
+    def get_non_selected_fields(self):
+        dimension = {"ga:browser","ga:operatingSystem","ga:flashVersion","ga:javaEnabled","ga:screenColors",
+                     "ga:screenResolution","ga:hostname","ga:date","ga:year","ga:month","ga:hour"}
+        selected_fields = set(random.sample(dimension, 9))  # randomly select 9 fields from dimension set
+        non_selected_fields = dimension - selected_fields
+        return non_selected_fields
```
Comment (RushT007) [inline tests/test_google_analytics_dynamic_dimension_metric.py:20]: As we discussed, we need generic implementation which will work with all the streams. 
Changes requested:
- include 2 parameters number of dimensions and metrics allowed
- write 2 more functions for random dimension and metric selection
- create class method which will validate the random dimension and metric combination to avoid testing invalid combinations.

File: tests/test_google_analytics_dynamic_dimension_metric.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+                messages = synced_records.get(stream)
+                # collect actual values
+                actual_all_keys = set()
+                for message in messages['messages']:
+                    if message['action'] == 'upsert':
+                        actual_all_keys.update(message['data'].keys())
+
+                # verify all the selected fields for each Audience Technology are replicated
+                self.assertSetEqual(expected_all_keys, actual_all_keys)
```
Comment (RushT007) [inline tests/test_google_analytics_dynamic_dimension_metric.py:87]: We need to iterate same test multiple times to have maximum field coverage. For this consider 30 mins as maximum execution time and plan the iterations.

File: tests/test_google_analytics_dynamic_dimension_metric.py (hunk 4 above)
Comment (RushT007) [inline tests/test_google_analytics_dynamic_dimension_metric.py:20]: As we discussed, we need generic implementation which will work with all the streams. 
Changes requested:
- include 2 parameters number of dimensions and metrics allowed
- write 2 more functions for random dimension and metric selection
- create class method which will validate the random dimension and metric combination to avoid testing invalid combinations.
- Define appropriate data structures

**Review by RushT007 — CHANGES_REQUESTED** [review]

**Review by RushT007 — APPROVED** [review]

File: tests/test_google_analytics_automatic_fields.py [hunk 6]
```diff
@@ -64,3 +64,11 @@ def test_run(self):
                 # Verify that only the automatic fields are sent to the target
                 for actual_keys in record_messages_keys:
                     self.assertSetEqual(expected_keys, actual_keys)
+
+                # Verify we did not duplicate any records across pages
+                records_pks_set = {tuple([message.get('data').get(primary_key) for primary_key in expected_keys])
+                                          for message in data.get('messages')}
+                records_pks_list = [tuple([message.get('data').get(primary_key) for primary_key in expected_keys])
+                                           for message in data.get('messages')]
```
Comment (RushT007) [inline tests/test_google_analytics_automatic_fields.py:72]: Suggestion, this code block repeats same statement twice and could be simplified as below.

e.g.

```
records_pks_list = [tuple([message.get('data').get(primary_key) for primary_key in expected_keys])
                                           for message in data.get('messages')]

self.assertCountEqual(records_pks_set, set(records_pks_list), msg="We have duplicate records for {}".format(stream))
```

Comment (umangagrawal-crest) [reply tests/test_google_analytics_automatic_fields.py]: Done

File: tests/test_google_analytics_dynamic_dimension_metric.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
+import tap_tester.connections as connections
```
Comment (kspeer825) [inline tests/test_google_analytics_dynamic_dimension_metric.py:1]: I think this test is a better implementation of the all_fields test for this tap. I don't think both are needed. This one should remain but the all fields test should be removed.

**Review by kspeer825 — CHANGES_REQUESTED** [review]

File: tests/test_google_analytics_dynamic_dimension_metric.py (hunk 7 above)
Comment (umangagrawal-crest) [reply tests/test_google_analytics_dynamic_dimension_metric.py:1]: @kspeer825 As per your comment removed all fields test cases.

File: tests/test_google_analytics_dynamic_dimension_metric.py (hunk 5 above)
Comment (umangagrawal-crest) [reply tests/test_google_analytics_dynamic_dimension_metric.py]: Ran the tests 5 times for the streams in which dynamically fields are selected.

File: tests/test_google_analytics_dynamic_dimension_metric.py (hunk 4 above)
Comment (umangagrawal-crest) [reply tests/test_google_analytics_dynamic_dimension_metric.py]: Created one function which will select dimensions and metrics randomly. Also, created one function which will validate the selected fields
